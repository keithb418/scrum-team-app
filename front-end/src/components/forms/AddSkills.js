import React, { Component } from "react";
import { FormGroup, FormControl, InputGroup, ControlLabel } from "react-bootstrap";
import FontAwesome from "react-fontawesome";
import PropTypes from "prop-types";

export default class AddSkills extends Component {
  constructor (props) {
    super(props);
    this.state = { 
      skills: this.props.skills || [],
      skill: '' 
    };
  }

  onAdd = () => { 
    this.setState(({skills, skill}) => ({ skills: skills.concat(skill) }));
    this.props.onChange(this.state.skills);
    this.resetInput();
  }

  onRemove = (e) => {
    let skillToRemove = e.target.parentNode.innerText;
    this.setState(({skills}) => ({ skills: skills.filter(skill => skill != skillToRemove) }));
    this.props.onChange(this.state.skills);
  }

  getInput = (e) => {
    const value = e.target.value;
    const name = e.target.name;

    this.setState({
      [name]: value
    });
  }

  resetInput = () => {
    this.setState(() => ({ skill: '' }));
  }

  render () {
    let skillsToAdd = this.state.skills;
    return (
      <div className="add-skills">
        <FormGroup controlId={this.props.id} style={{ width: "200px" }}>
          <ControlLabel>Skills</ControlLabel>
          <InputGroup>
            <FormControl name="skill" type="text" placeholder="Add Skills" value={this.state.skill} onChange={this.getInput}/>
            <InputGroup.Addon>
              <FontAwesome name="plus-circle" className="left-spacer add-skills-icon" style={{ cursor: "pointer" }} onClick={this.onAdd} />
            </InputGroup.Addon>
          </InputGroup>
        </FormGroup>

        {skillsToAdd.map((skill, index) => {
          return (
            <div style={{ display: "inline-block", marginRight: "15px" }} className="skill" key={index}>{skill}
              <FontAwesome name="minus-circle" style={{ cursor: "pointer" }} onClick={this.onRemove} />
            </div>
          );
        })}

      </div>
    );
  }
}

AddSkills.propTypes = {
  id: PropTypes.string,
  onChange: PropTypes.func
};