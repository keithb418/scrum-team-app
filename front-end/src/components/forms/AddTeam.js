import React from "react";
import { connect } from "react-redux";
import { FormGroup, ControlLabel, FormControl, Button } from "react-bootstrap";

let teamId = 200;

class AddTeam extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      team: "",
      title: "Add a New Team"
    };
    this.onInputChange = this.onInputChange.bind(this);
  }

  render() {
    return (
      <div class="row">
        <div class="col-md-8">
          <h3>{this.state.title}</h3>
          <Form>
            <FormGroup controlId="team">
              <FormControl
                type="text"
                placeholder="Enter team name"
                onChange={this.onInputChange}
              />
            </FormGroup>
            <Button onClick={() => dispatch({
              type: "ADD_TEAM",
              team: {
                "_id": `${teamId++}team`,
                "team": this.state.team
              }
            })}>Add Team
            </Button>
          </Form>
        </div>
      </div>
    );
  };

  onInputChange() {
    const target = event.target;
    const value = target.value;
    const name = target.name;

    this.setState({
      [name]: value
    });
  }
};

AddTeam = connect((state, ownProps) => {
  return {
    name: state.name
  }
})(AddTeam);


export default AddTeam;