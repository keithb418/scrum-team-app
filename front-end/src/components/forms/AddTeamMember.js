import React, { Component } from "react";
import { connect } from "react-redux";
import TeamMemberForm from "./TeamMemberForm";
import { handleCreateTeamMember } from "../../actions/teamMembers";
import TeamMember from "../TeamMember";

export class AddTeamMember extends Component {
  constructor (props) {
    super(props);
  }

  onSubmit = (teamMember) => {
    this.props.onSubmit(teamMember);
    this.props.history.push('/');
  }

  render() {
    return (
      <TeamMemberForm
        title="Add Team Member"
        teams={this.props.teams}
        roles={this.props.roles}
        onSubmit={this.onSubmit}
      />
    )
  }
}

const mapStateToProps = ({ teams: { teams }, roles } ) => ({
    teams,
    roles
})

const mapDispatchToProps = (dispatch) => ({
  onSubmit: (teamMember) => dispatch(handleCreateTeamMember(teamMember))
})

export default connect(mapStateToProps, mapDispatchToProps)(AddTeamMember);