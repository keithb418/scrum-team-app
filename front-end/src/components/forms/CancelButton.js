import React from "react";
import { Button } from "react-bootstrap";
import { Link } from "react-router-dom";

const CancelButton = () => (
  <Link to="/">
    <Button>Cancel</Button>
  </Link>
);


export default CancelButton;
