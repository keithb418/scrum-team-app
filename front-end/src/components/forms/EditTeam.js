import React, { Component } from "react";
import { connect } from "react-redux";
import TeamForm from "./TeamForm";
import Loading from "../Loading";
import { handleUpdateTeam, handleFetchTeam } from "../../actions/teams";
import { handleFetchTeamMember } from "../../actions/teamMembers";

export class EditTeam extends Component {
  componentDidMount() {
    this.props.fetchTeam(this.props.match.params.id)
  }

  onSubmit = (team) => {
    const _id = this.props.team._id;
    this.props.editTeam({
      _id,
      ...team
    });
    this.props.history.push('/');
  }

  render() {
    return this.props.isFetching ?
      <Loading /> :
      <TeamForm
        title="Edit Team"
        team={this.props.team}
        onSubmit={this.onSubmit}
      />
  }
}

const mapStateToProps = ({ teams: { team }, fetch: { isFetching } }) => ({
    team,
    isFetching
});

const mapDispatchToProps = (dispatch) => ({
  fetchTeam: (id) => dispatch(handleFetchTeam(id)),
  editTeam: (team) => dispatch(handleUpdateTeam(team))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditTeam);