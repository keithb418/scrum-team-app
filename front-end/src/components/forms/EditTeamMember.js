import React, { Component } from "react";
import { connect } from "react-redux";
import TeamMemberForm from "./TeamMemberForm";
import { fetchProfileData } from '../../actions/shared';
import { handleUpdateTeamMember } from "../../actions/teamMembers";
import TeamMember from "../TeamMember";
import Loading from "../Loading";

export class EditTeamMember extends Component {
  constructor (props) {
    super(props);
  }

  componentDidMount() {
    this.props.fetchProfileData(this.props.match.params.id)
  }

  onSubmit = (teamMember) => {
    const _id = this.props.teamMember._id;
    this.props.editTeamMember({
      _id,
      ...teamMember
    });
    this.props.history.push('/');
  }
  
  render() {
    return this.props.isFetching ?
      <Loading /> :
      <TeamMemberForm
        title="Edit Team"
        teams={this.props.teams}
        roles={this.props.roles}
        teamMember={this.props.teamMember}
        onSubmit={this.onSubmit}
      />
  }
}

const mapStateToProps = ({ teamMembers: { teamMember }, teams: { teams }, roles }) => ({
  teamMember,
  teams,
  roles,
})

const mapDispatchToProps = (dispatch) => ({
  fetchProfileData: (id) => dispatch(fetchProfileData(id)),
  editTeamMember: (teamMember) => dispatch(handleUpdateTeamMember(teamMember))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditTeamMember);