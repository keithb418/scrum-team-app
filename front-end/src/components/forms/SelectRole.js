import React from "react";
import { FormGroup, ControlLabel, FormControl } from "react-bootstrap";
import PropTypes from "prop-types";

const SelectRole = ({ roles = [], selected = "", onSelect }) => {
  return (
    <FormGroup controlId="roleSelect">
      <ControlLabel>Role</ControlLabel>
      <FormControl
        name="role"
        defaultValue={selected}
        componentClass="select"
        placeholder="Select Role"
        onChange={ onSelect }>

        <option>Select Role</option>

        {roles.map(role =>
          <option
            value={role.name}
            key={role._id}>{role.name}</option>
        )}

      </FormControl>
    </FormGroup>
  );
};

SelectRole.propTypes = {
  onSelect: PropTypes.func,
  roles: PropTypes.array,
};

export default SelectRole;