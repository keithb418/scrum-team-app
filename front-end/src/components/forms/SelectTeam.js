import React from "react";
import { FormGroup, ControlLabel, FormControl } from "react-bootstrap";
import PropTypes from "prop-types";

const SelectTeam = ({ teams = [], selected = "", onSelect }) => {
  return (
    <FormGroup controlId="teamSelect">
      <ControlLabel>Team<sup><i className="fa fa-asterisk required"></i></sup></ControlLabel>
      <FormControl
        name="team"
        defaultValue={selected}
      	componentClass="select"
      	placeholder="Select Team"
      	onChange={onSelect} >

        <option>Select Team</option>

        {teams.map(team =>
          <option
            value={team._id}
            key={team._id}>{team.name}</option>
        )}

      </FormControl>
    </FormGroup>
  );
};

SelectTeam.propTypes = {
  onSelect: PropTypes.func,
  teams: PropTypes.array,
};

export default SelectTeam;