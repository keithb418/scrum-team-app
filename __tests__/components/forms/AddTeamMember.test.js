import React from 'react';
import { shallow } from 'enzyme';
import { AddTeamMember } from '../../../front-end/src/components/forms/AddTeamMember';
import teamMemberData from '../../fixtures/teamMembers';
import teamData from '../../fixtures/teams';
import roleData from '../../fixtures/roles';

let onSubmit, history, wrapper;
beforeEach(() => {
    onSubmit = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(
        <AddTeamMember 
            onSubmit={onSubmit} 
            history={history} 
            teams={teamData}
            roles={roleData}/>
    );
});

test('should render AddTeamMember component correctly', () => {
    expect(wrapper).toMatchSnapshot();
});
  
test('should handle onSubmit', () => {
    wrapper.find('TeamMemberForm').prop('onSubmit')(teamMemberData[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(onSubmit).toHaveBeenLastCalledWith(teamMemberData[1]);
});