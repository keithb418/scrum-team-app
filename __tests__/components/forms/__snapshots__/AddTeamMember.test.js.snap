// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddTeamMember component correctly 1`] = `
<TeamMemberForm
  onSubmit={[Function]}
  roles={
    Array [
      Object {
        "_id": "1",
        "name": "Front-End Developer",
      },
      Object {
        "_id": "2",
        "name": "Back-End Developer",
      },
      Object {
        "_id": "3",
        "name": "Full-Stack Developer",
      },
      Object {
        "_id": "4",
        "name": "Manual Tester",
      },
      Object {
        "_id": "5",
        "name": "Automation Tester",
      },
    ]
  }
  teams={
    Array [
      Object {
        "_id": "1",
        "name": "Team 1",
      },
      Object {
        "_id": "2",
        "name": "Team 2",
      },
    ]
  }
  title="Add Team Member"
/>
`;
